src/Microsoft.SymbolStore/ChecksumValidator.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.FileFormats.PE;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.SymbolStore
+{
+    class ChecksumValidator
+    {
+        private const string pdbStreamName = "#Pdb";
+        private const uint pdbIdSize = 20;
+
+        internal static void Validate(ITracer tracer, Stream pdbStream, IEnumerable<PdbChecksum> pdbChecksums)
+        {
+            using (MemoryStream ms = new MemoryStream())
+            {
+                pdbStream.CopyTo(ms);
+                byte[] bytes = ms.ToArray();
+                uint offset = 0;
+
+                try
+                {
+                    offset = GetPdbStreamOffset(pdbStream);
+                }
+                catch (Exception ex)
+                {
+                    tracer.Error(ex.Message);
+                    throw;
+                }
+
+                for (int i = 0; i <= pdbIdSize; i++)
+                {
+                    bytes[i + offset] = 0;
+                }
+
+                foreach (var checksum in pdbChecksums)
+                {
+                    tracer.Information($"Testing checksum: {checksum}");
+
+                    var algorithm = HashAlgorithm.Create(checksum.AlgorithmName);
+                    if (algorithm != null)
+                    {
+                        var hash = algorithm.ComputeHash(bytes);
+                        if (hash.SequenceEqual(checksum.Checksum))
+                        {
+                            // If any of the checksums are OK, we're good
+                            tracer.Information($"Found checksum match {checksum}");
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        throw new ArgumentException("Unknown hash algorithm");
+                    }
+                }
+            }
+            throw new InvalidChecksumException("PDB checksum mismatch");
```

Comment:
To make sure I understand, the idea here is that the InvalidChecksumException makes it all the way back to the client, is that right? If so, might we worth calling that out in the doc comments for the public API [here](https://github.com/dotnet/symstore/blob/18f84ad901fbf454dccc846be2f0cadf1a53e25a/src/Microsoft.SymbolStore/SymbolStores/SymbolStore.cs#L39)